include/deal.II/base/timer_tree.h

```diff
@@ -0,0 +1,423 @@
+
+        std::shared_ptr<TimerTree> new_tree(new TimerTree());
+        new_tree->copy_from(sub_tree);
+        if (!new_name.empty())
+          new_tree->id = new_name;
+
+        // Make sure that the new tree does not already exist
+        for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+          {
+            AssertThrow(
+              new_tree->id != (*it)->id,
+              ExcMessage(
+                "Subtree can not be inserted since the tree already contains a subtree with the same ID."));
+          }
+
+        sub_trees.push_back(new_tree);
+      }
+  }
+
+  /**
+   * Prints wall time of all items of a tree without an analysis of
+   * the relative share of the children.
+   */
+  void
+  print_plain(ConditionalOStream const &pcout) const
+  {
+    unsigned int const length = get_length();
+
+    pcout << std::endl;
+
+    do_print_plain(pcout, 0, length);
+  }
+
+  /**
+   * This is the actual function of interest of this class, i.e., an
+   * analysis of wall times with a hierarchical formatting of results.
+   * Relative wall times are printed for all children of a sub-tree if
+   * a wall time has been set for the parent of these children. In this
+   * case, an additional item `other` is created in order to give insights
+   * to which extent the code has been covered with timers and to which
+   * extend time is spent in other code paths that are currently not
+   * covered by timers. The parameter @p level specifies the level for which
+   * to print results, where a value of 0 corresponds to the high level module.
+   */
+  void
+  print_level(ConditionalOStream const &pcout, unsigned int const level) const
+  {
+    unsigned int const length = get_length();
+
+    pcout << std::endl;
+
+    do_print_level(pcout, level, 0, length);
+  }
+
+private:
+  /*
+   * Copy function.
+   */
+  void
+  copy_from(std::shared_ptr<TimerTree const> const other)
+  {
+    *this = *other;
+  }
+
+  /*
+   * This function erases the first entry of the vector.
+   */
+  std::vector<std::string>
+  erase_first(std::vector<std::string> const &in) const
+  {
+    AssertThrow(in.size() > 0, ExcMessage("empty name."));
+
+    std::vector<std::string> out(in);
+    out.erase(out.begin());
+
+    return out;
+  }
+
+  /*
+   * Returns average wall-time over all MPI processes for the root of this tree.
+   */
+  double
+  get_average_wall_time() const
+  {
+    Utilities::MPI::MinMaxAvg time_data =
+      Utilities::MPI::min_max_avg(data->wall_time, MPI_COMM_WORLD);
+    return time_data.avg;
+  }
+
+  /*
+   * print functions
+   */
+  unsigned int
+  get_length() const
+  {
+    unsigned int length = id.length();
+
+    for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+      {
+        length = std::max(length, (*it)->get_length() + offset_per_level);
+      }
+
+    return length;
+  }
+
+  void
+  do_print_plain(ConditionalOStream const &pcout,
```

Comment:
All of these private member functions, as well as the member variables and internal structs, should also be documented.

Reply:
done here https://github.com/dealii/dealii/pull/11939/commits/78670552974857ec20150fcc0a7b464bfdbbd087 and https://github.com/dealii/dealii/pull/11939/commits/787cf55e96112e43d7d042a6e1cd6f2d01b912d9